### DISCUSSION TITLE
Help moving from Elsa 1.5 to Elsa 2.0

### DISCUSSION THREAD

**Discussion by Alexplose** [elsa-workflows/elsa-core#1232] [Q&A]
Hi,

I'm my project is now targeting .net5 so I'm also moving to Elsa 2.0, a lot of thing have changed since, and I'm a bit lost.

I'll make a list of those changes to get some help here and maybe other that are lost too.

First : 

**Properties that were defined as :** 

`       [ActivityProperty(Hint = "The status to set.")]

        public WorkflowExpression<PeriodStatus> Status
        {
            get => GetState<WorkflowExpression<PeriodStatus>>();
            set => SetState(value);
        }
`
Should be changed to 

`       [ActivityProperty(Hint = "The status to set.")]

        public PeriodStatus Status
       {
            get => GetState<PeriodStatus>();
            set => SetState(value);
        }
`

And the call to  `var user = await context.EvaluateAsync(PeriodStatus, cancellationToken);` should be removed use directly the property.

**The ActivityDefinitionAttribute has disappeared.**
`[ActivityDefinition(`
And is now ActivityAttribute
`[Activity(`

Icon property is not available anymore.

`TriggerSignalAsync` is now in the `ISignaler` interface.

Halt OutcomeResult has been replaced with Suspend

**Now the questions :** 

I was using the `AddLogEntry` method to add some custom executionlog, is there any way to add logs and see those in the dashboard with Elsa 2.0 ?

`IWorkflowInvoker` has been replaced with `IWorkflowRunner `?

How do you `workflowRegistry.GetWorkflowDefinitionAsync`  now ?
And the same for : 
`workflowInstanceStore.GetByCorrelationIdAsync`
`_workflowDefinitionStore.GetByIdAsync(instance.DefinitionId, VersionOptions.SpecificVersion(instance.Version));`
`_workflowInstanceStore.ListByStatusAsync`

Also is there any migration tips for existing Elsa 1.5 workflows & instances ?

I'm sorry if I missed those explanations in the docs, but I didn't find anything.

Help appreciated.

**Comment by Alexplose** [comment]
Another finding, the ExecutionLog is not searchable in the version published as nugget packages, but the IWorkflowExecutionLog interface has 
`Task<WorkflowExecutionLogRecord?> FindEntryAsync(ISpecification<WorkflowExecutionLogRecord> specification, CancellationToken cancellationToken = default);`

 in the master branch.

How can I access ExecutionLog ?

**Reply by sfmskywalker** [reply]
There are two ways to access it: via the higher-level service `IWorkflowExecutionLog`. Right now it allows you to only select a single record, but if you need a method to select more than just one, then please create a feature request. It makes sense for this service to have it.

The second way is to access the `IWorkflowExecutionLogStore` service, which should give you everything you need to query the execution log entries for a given workflow instance.

**Comment by Alexplose** [comment]
I'm unable to get Elsa migration to update my existing database.
Even when defining MigrationsHistoryTable

`  services.AddElsa(elsa => {
                elsa.UseEntityFrameworkPersistence(
configure => configure.UseSqlServer(configuration.GetConnectionString("Db"), c=>c.MigrationsHistoryTable("_EFMigrationsHistoryElsa", "Elsa")).EnableDetailedErrors());`

Migration table has been created but still empty and no tables are created.

**Reply by sfmskywalker** [reply]
If you existing database contains Elsa 1 tables, they will not be migrated to thee Elsa 2 schema.

Elsa 2 contains a lot of breaking changes making it less than trivial to automatically migrate.
To migrate from 1 to 2, it is best to start with a clean database (or use the same database, Elsa 2 will create its tables under the "Elsa" schema, so there won't be any conflicts with table names etc).

Elsa 1 workflows need to be manually re-created step-by-step into an Elsa 2 workflow. It sucks, but writing a migration tool is tricky and too time-consuming. It might be worthwhile if you have many large workflows that would take longer than a few months to migrate (this is the amount of time writing a migration tool that works perfectly might take, if not longer, to discover and handle all the changes).

If you are generating new tables for Elsa 2 using the migrations, then that should just work provided that you either uses Elsa's version of the `UseSqlServer` extension method, or the Microsoft one, but then you need to manually specify the migrations assembly as [done here](https://github.com/elsa-workflows/elsa-core/blob/master/src/persistence/Elsa.Persistence.EntityFramework/Elsa.Persistence.EntityFramework.SqlServer/DbContextOptionsBuilderExtensions.cs#L11-L14).

**Comment by sfmskywalker** [comment]
Hi @Alexplose ,

> I was using the AddLogEntry method to add some custom executionlog, is there any way to add logs and see those in the dashboard with Elsa 2.0 ?

Yes, execution log is now stored separately from the workflow instance for performance reasons (some workflows have a very long lifespan with a lot of iterations).

You can access this via `IWorkflowExecutionLog`.

The log entries are visible from the dashboard when you go the the workflow instance viewer screen. They are displayed to the right on the Workflow Journal panel.

> How do you workflowRegistry.GetWorkflowDefinitionAsync now ?

To get a single **workflow blueprint**, you can use `IWorkflowRegistry.GetAsync`. Example:

```csharp
_workflowRegistry.GetAsync("MyWorkflowDefinitionId", version: VersionOptions.Published);
```

If you want to get a single **workflow definition**, use `IWorkflowDefinitionStore`.

> Also is there any migration tips for existing Elsa 1.5 workflows & instances ?

There are none documented I'm afraid. The discussion you started here is the closest to it. But I will take it as a basis and create a documentation page where we can collect tips and tricks for migrating from v1 to v2.

**Reply by sfmskywalker** [reply]
I started a documentation page related to migrating from 1.x to 2.0 [here](https://elsa-workflows.github.io/elsa-core/docs/next/upgrading/upgrading-v1.x-to-v2.0).

**Comment by Alexplose** [comment]
Here is my quick & dirty Elsa 1.5 to 2.0 converter, it's far from being perfect, and will allow you at least to recreate activity and connection, to avoid rebuilding all the workflow from scratch, some variables are sets, others are not.

[ElsaConverter](https://github.com/Alexplose/ElsaConverter)

Download workflow as json from you Elsa 1.5 instance, and save it with filename matching XXX.workflow_name.workflow_definition_id.json 
Change source files to your needs.

**Reply by sfmskywalker** [reply]
Nice! Thanks for sharing this.
